File: 19th/코딩테스트 기초반/week4/8611번_팰린드롬 숫자.cpp
#include <iostream>
#include <string>

using namespace std;

// 10^1000이기에 string으로 계산해야 함
// string용 나눗셈 함수 생성
// divide 함수는 몫과 나머지를 반환
pair<string, int> divide(string num, int base) { 
    string res = "";
    int remainder = 0; // 현재 나머지

    for (int i = 0; i < num.size(); i++) {
        // 제일 앞 자리(높은 자리수) 수부터 나누고, 만약 이전에 나누지 못한 수가 있을 경우 10을 곱한 후 현재 자리의 수를 더함
        int current = remainder * 10 + (num[i] - '0');
        if (!res.empty() || current / base != 0) res += char('0' + current / base); // 몫의 맨 앞에 0이 들어가는 것을 방지
        remainder = current % base; // 현재의 나머지 갱신
    }

    if (res.empty()) res = "0"; // 몫이 없을 경우 0 반환
    return {res, remainder};
}

// 진법 변환 함수
string convert(string num, int base) {
    string res = "";

    while (num != "0") {
        pair<string, int> divided_value = divide(num, base);
        res = char('0' + divided_value.second) + res; // 변환하고자 하는 진법(base)으로 나눈 나머지를 낮은 자리부터 저장
        num = divided_value.first; // 변환해야 할 수
    }

    return res;
}

int main() {
    string num;
    cin >> num;

    bool is_palindrome = false; // 팰린드롬인 경우가 있는지 판단하기 위함
    for (int i = 2; i <= 10; i++) {
        string tmp = convert(num, i);
        bool is_passible = true; // 현재의 진법으로 변환된 값이 팰린드롬인지 판단하기 위함

        int size = tmp.size();
        for (int j = 0; j <= size / 2; j++) {
            // 중앙을 기준으로 같은 거리에 있는 값들이 다를 경우 팰린드롬이 아님
            if (tmp[j] != tmp[size - j - 1]) {
                is_passible = false;
                break;
            }
        }

        if (is_passible) { // 팰린드롬일 경우
            is_palindrome = true; // 팰린드롬인 경우가 있기에 true로 갱신
            cout << i << " " << tmp << "\n";
        }
    }

    if (!is_palindrome) { // 팰린드롬인 경우가 없는 경우
        cout << "NIE";
    }
}

File: 19th/코딩테스트 기초반/week4/8611번_팰린드롬 숫자.py
# num을 base(진수)로 변환하여 문자열로 반환하는 함수
def to_base_n(num, base):
    digits = "0123456789"  # 진법 변환 시 사용할 숫자들
    if num == 0:
        return "0"  # 0은 모든 진법에서 0
    result = ''
    while num > 0:
        result = digits[num % base] + result  # 나머지를 앞에 붙여 나감
        num //= base
    return result

# 팰린드롬(회문)인지 확인하는 함수: 문자열이 앞뒤로 같으면 True
def check_palindrome(s):
    return s == s[::-1]

# 입력 받기
n = int(input())
flag = 1  # 팰린드롬을 발견했는지 여부를 체크

# 2진수부터 9진수까지 확인
for i in range(2, 10):
    num = to_base_n(n, i)  # n을 i진수로 변환
    if check_palindrome(num):  # 팰린드롬인지 확인
        print(i, num)  # 진수와 변환된 수 출력
        flag = 0  # 팰린드롬 발견했으므로 flag 0으로 설정

# 10진수로도 팰린드롬인지 확인
if check_palindrome(str(n)):
    flag = 0
    print(10, n)
elif flag:
    print('NIE')  # 어떤 진수에서도 팰린드롬이 아니면 NIE 출력
